src/Juvix/Backends/Michelson/Parameterisation.hs

```diff
@@ -37,6 +44,52 @@ constType v =
 arity :: PrimVal -> Int
 arity = pred . length . typeOf
 
+applyProper ::
+  Prim.Take PrimTy PrimVal ->
+  [Prim.Take PrimTy PrimVal] ->
+  Either
+    (CoreTypes.PipelineError PrimTy PrimVal)
+    (Prim.Return PrimTy PrimVal)
+applyProper fun args =
+  case Prim.term fun of
+    Constant _i ->
+      case length args of
+        0 ->
+          Right (Prim.Return (Prim.term fun))
+        _x ->
+          Left (CoreTypes.TypecheckerError "Applied a constant to argument")
+    Inst instruction ->
+      let inst = Instructions.toNumArgs instruction
+       in case inst `compare` fromIntegral (length args) of
+            -- we should never take more arguments than primitve could handle
+            GT ->
+              "applied too many arguments to a Michelson Primitive"
+                |> CoreTypes.TypecheckerError
+                |> Left
+            LT ->
+              inst - fromIntegral (length args)
+                |> Prim.Cont fun args
+                |> Right
+            -- we have exactly the right number of arguments, call the interpreter!
+            EQ ->
+              let newTerm =
+                    Run.applyPrimOnArgs (Prim.toAnn fun) (Prim.toAnn <$> args)
+                  -- TODO ∷ do something with the logs!?
+                  (compd, _log) = Compilation.compileExpr newTerm
+               in case compd >>= Interpreter.dummyInterpret of
+                    Right x ->
+                      Constant x
+                        |> Prim.Return
+                        |> Right
+                    -- TODO :: promote this error
```

Comment:
I believe that this failure indicates a mistake in the typechecker, which should have been run already by this stage

Reply:
I'd still like to reserve the errors of the compiler itself, and so an Either over Maybe is nicer?

Reply:
This is now promoted, however I changed the 

Juvix.Core.Types.TermAssignment